File: docs/snippets/all/descriptors/descr_custom_archetype.py
#!/usr/bin/env python3

from __future__ import annotations

from typing import TYPE_CHECKING, Any

import dalaran as dl  # pip install dalaran-sdk

if TYPE_CHECKING:
    import numpy.typing as npt


class CustomPoints3D(dl.AsComponents):  # type: ignore[misc]
    def __init__(
        self: Any, positions: npt.ArrayLike, colors: npt.ArrayLike
    ) -> None:
        self.positions = dl.components.Position3DBatch(positions).described(
            dl.ComponentDescriptor(
                "user.CustomPoints3D:custom_positions",
                archetype="user.CustomPoints3D",
                component_type="user.CustomPosition3D",
            ),
        )
        self.colors = dl.components.ColorBatch(colors).described(
            dl.ComponentDescriptor("user.CustomPoints3D:colors").with_overrides(
                archetype="user.CustomPoints3D",
                component_type=dl.components.ColorBatch._COMPONENT_TYPE,
            )
        )

    def as_component_batches(self) -> list[dl.DescribedComponentBatch]:
        return [self.positions, self.colors]


dl.init("dalaran_example_descriptors_custom_archetype")
dl.spawn()

dl.log("data", CustomPoints3D([[1, 2, 3]], [0xFF00FFFF]), static=True)

# The tags are indirectly checked by the Rust version (have a look over there
# for more info).

File: docs/snippets/all/descriptors/descr_custom_archetype.cpp
#include <dalaran.hpp>
#include <vector>

struct CustomPosition3D {
    dalaran::Position3D position;
};

template <>
struct dalaran::Loggable<CustomPosition3D> {
    static constexpr ComponentDescriptor Descriptor = "user.CustomPosition3D";

    static const std::shared_ptr<arrow::DataType>& arrow_datatype() {
        return dalaran::Loggable<dalaran::Position3D>::arrow_datatype();
    }

    // TODO(#4257) should take a dalaran::Collection instead of pointer and size.
    static dalaran::Result<std::shared_ptr<arrow::Array>> to_arrow(
        const CustomPosition3D* instances, size_t num_instances
    ) {
        return dalaran::Loggable<dalaran::Position3D>::to_arrow(
            reinterpret_cast<const dalaran::Position3D*>(instances),
            num_instances
        );
    }
};

/// A custom archetype that extends Dalaran's builtin `dalaran::Points3D` archetype with a custom component.
struct CustomPoints3D {
    dalaran::Collection<CustomPosition3D> positions;
    std::optional<dalaran::Collection<dalaran::Color>> colors;
};

template <>
struct dalaran::AsComponents<CustomPoints3D> {
    static Result<dalaran::Collection<ComponentBatch>> as_batches(
        const CustomPoints3D& archetype
    ) {
        std::vector<dalaran::ComponentBatch> batches;

        auto positions_descr = dalaran::ComponentDescriptor(
            "user.CustomPoints3D",
            "user.CustomPoints3D:custom_positions",
            "user.CustomPosition3D"
        );
        batches.push_back(
            ComponentBatch::from_loggable(archetype.positions, positions_descr)
                .value_or_throw()
        );

        if (archetype.colors) {
            auto colors_descr =
                dalaran::ComponentDescriptor("user.CustomPoints3D:colors")
                    .with_archetype("user.CustomPoints3D")
                    .with_component_type(
                        dalaran::Loggable<dalaran::Color>::ComponentType
                    );
            batches.push_back(
                ComponentBatch::from_loggable(archetype.colors, colors_descr)
                    .value_or_throw()
            );
        }

        return dalaran::take_ownership(std::move(batches));
    }
};

int main(int argc, char* argv[]) {
    const auto rec =
        dalaran::RecordingStream("dalaran_example_descriptors_custom_archetype");
    rec.spawn().exit_on_failure();

    rec.log_static(
        "data",
        CustomPoints3D{
            CustomPosition3D{{1.0f, 2.0f, 3.0f}},
            dalaran::Color(0xFF00FFFF)
        }
    );

    // The tags are indirectly checked by the Rust version (have a look over there for more info).
}
